File: CS50X/Week 1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool validCreditCard(long cardNumber);
string creditCardType(long cardNumber);

int main(void)
{
    long cardNumber;

    // Prompt the user for a credit card number
    do
    {
        cardNumber = get_long("Number: ");
    }
    while (cardNumber <= 0);

    // Validate the credit card number using Luhn's Algorithm
    if (validCreditCard(cardNumber))
    {
        printf("%s\n", creditCardType(cardNumber));
    }
    else
    {
        printf("INVALID\n");
    }
}

// Function to check if a credit card number is valid using Luhn's Algorithm
bool validCreditCard(long cardNumber)
{
    int sum = 0;
    int count = 0;

    while (cardNumber > 0)
    {
        // Get the last digit
        int digit = cardNumber % 10;

        if (count % 2 == 0)
        {
            // If the digit is at an even position, add it directly
            sum += digit;
        }
        else
        {
            // If the digit is at an odd position, multiply it by 2 and add the digits of the
            // product
            int temp = 2 * digit;
            int x = 0;
            while (temp > 0)
            {
                x += temp % 10;
                temp /= 10;
            }

            sum += x;
        }

        // Move to the next digit
        cardNumber /= 10;
        count++;
    }

    // Check if the total sum's last digit is 0
    return sum % 10 == 0;
}

string creditCardType(long cardNumber)
{
    string cardType = "";
    long temp = cardNumber;

    // Extract the first digit(s) of the card number
    while (cardNumber >= 100)
    {
        cardNumber /= 10;
    }

    int numDigits = 0;
    while (temp != 0)
    {
        temp /= 10;
        numDigits++;
    }

    // Determine the card type based on the first digit(s)
    if ((numDigits == 13 || numDigits == 16) && (cardNumber >= 40 && cardNumber <= 49))
    {
        cardType = "VISA";
    }
    else if (numDigits == 16 && (cardNumber >= 51 && cardNumber <= 55))
    {
        cardType = "MASTERCARD";
    }
    else if (numDigits == 15 && (cardNumber == 34 || cardNumber == 37))
    {
        cardType = "AMEX";
    }
    else
    {
        cardType = "INVALID";
    }

    return cardType;
}

File: CS50X/Week 6/sentimental-credit/credit.py
import re


def get_card_type(number):
    # Return the type of credit card based on the card number.
    if re.match(r'^3[47][0-9]{13}$', number):
        return "AMEX"
    elif re.match(r'^5[1-5][0-9]{14}$', number):
        return "MASTERCARD"
    elif re.match(r'^4[0-9]{12}(?:[0-9]{3})?$', number):
        return "VISA"
    else:
        return "INVALID"


def luhnAlgorithm(number):
    # Use the Luhn algorithm to validate the credit card number.
    total = 0
    num_digits = len(number)
    parity = num_digits % 2

    for i, digit in enumerate(number):
        digit = int(digit)
        if i % 2 == parity:
            digit *= 2
        if digit > 9:
            digit -= 9
        total += digit

    return total % 10 == 0


def main():
    card_number = input("Number: ").strip()

    if not card_number.isdigit():
        print("INVALID")
        return

    card_type = get_card_type(card_number)

    if card_type != "INVALID" and luhnAlgorithm(card_number):
        print(card_type)
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
